File: ios/ZhtpAuth.swift
import Foundation
import Security

// MARK: - UHP v2 Authentication Implementation
//
// This file implements the post-handshake authentication layer for UHP v2.
// The handshake itself is performed by the Rust FFI layer (UhpHandshake.swift).
//
// ARCHITECTURE:
//   1. Handshake Phase (Rust Layer - UhpHandshake.swift)
//      - ClientHello: sends DID + Dilithium5 PK + Kyber1024 PK + challenge_nonce
//      - ServerHello: receives server identity, validates signature, extracts nonces
//      - ClientFinish: sends Kyber ciphertext, derives session keys
//      - Returns: sessionKey, sessionId (32 bytes), handshakeHash, peerDid
//
//   2. Session Derivation (Rust FFI result)
//      - sessionKey: 32-byte hybrid key (HKDF(kyberSharedSecret, classicalKey))
//      - sessionId: 32-byte HKDF(sessionKey, clientNonce || serverNonce)
//      - handshakeHash: 32-byte SHA3-256(transcript)
//
//   3. Authenticated Request Phase (iOS Layer - this file)
//      - Create AuthSession from handshake results
//      - For each request:
//        a. Increment monotonic counter (strict: no equal, no backwards)
//        b. Build canonical request: [method|pathLen|path|bodyLen|body|counter|sessionId]
//        c. MAC = HMAC-SHA3-256(macKey, canonicalRequest)
//        d. Send: [sessionId|counter|MAC] + request
//
// CRITICAL SECURITY PROPERTIES:
//   - Server signature verified in Rust layer (MITM protection)
//   - Timestamp skew validated in Rust layer (5-minute window, 300 seconds)
//   - MAC computed over EXACT canonical format (prevents tampering)
//   - Counter strictly monotonic (replay protection)
//   - Session ID exactly 32 bytes (protocol compliance)
//   - PQC hybrid key exchange mandatory (Kyber1024 + classical)
//
// SPECIFICATION REFERENCES:
//   - UHP v2 Handshake: lib-network/src/handshake/core.rs
//   - Session Keys: sdk-ts/src/quic/uhp_v2_handshake.ts:714-741
//   - MAC Computation: lib-network/src/protocols/types/session.rs:655-768

// MARK: - ZHTP Authentication (Public Mode)

/// Session state for authenticated connections
struct AuthSession {
    let sessionId: Data           // Session ID (16 or 32 bytes, per server)
    let macKey: Data              // [u8; 32] from HKDF derivation
    var sequence: UInt64          // monotonic counter, incremented per request
    let clientDid: String         // client identity
    let serverDid: String         // server identity
    let createdAt: Date           // session creation timestamp
    var lastActivity: Date        // last request timestamp

    /// Check if session is still valid
    /// - Not idle for > 5 minutes
    /// - Not older than 1 hour
    func isValid() -> Bool {
        let now = Date()

        // 5-minute idle timeout
        if now.timeIntervalSince(lastActivity) > 300 {
            return false
        }

        // 1-hour age limit
        if now.timeIntervalSince(createdAt) > 3600 {
            return false
        }

        return true
    }

    /// Update last activity timestamp
    mutating func touch() {
        lastActivity = Date()
    }

    /// Increment sequence counter for next request (wraps on overflow)
    /// UHP v2 requires strict monotonic increasing with no duplicates
    mutating func nextSequence() -> UInt64 {
        let current = sequence
        sequence = sequence &+ 1  // Wraps on overflow (unlikely in practice)
        return current
    }

    /// Validate that sequence never goes backwards (critical for replay protection)
    func validateSequenceNotRepeated(_ newSequence: UInt64) -> Bool {
        // Sequence must always be strictly increasing (never equal or less than previous)
        return newSequence > sequence
    }
}

func deriveMacKey(sessionKey: Data, handshakeHash: Data) throws -> Data {
    let info = "zhtp/v2/mac_key".data(using: .utf8) ?? Data()
    return try hkdfSha3(ikm: sessionKey, salt: handshakeHash, info: info, outputLength: 32)
}

/// Authentication context sent in ZHTP request header
struct AuthContext: Codable {
    let session_id: Data    // [u8; 32] - UHP v2 fixed size
    let client_did: String
    let sequence: UInt64
    let request_mac: Data   // [u8; 32] - HMAC-SHA3-256 output
}

// MARK: - Canonical Request Bytes (UHP v2)

/// Build canonical request bytes for MAC computation
///
/// Format (big-endian throughout, except signature is already BE):
/// [method: u8][pathLen: u16 BE][path bytes][bodyLen: u32 BE][body bytes][counter: u64 BE][sessionId: 32 bytes]
///
/// This exact format MUST be used for both MAC computation and verification.
/// Order is critical - any deviation will cause MAC mismatch.
func buildMacInput(
    method: ZhtpMethod,
    path: String,
    body: Data,
    counter: UInt64,
    sessionId: Data
) -> Data? {
    guard let methodByte = zhtpMethodByte(method) else {
        return nil
    }

    // Validate session ID is exactly 32 bytes (UHP v2 requirement)
    guard sessionId.count == 32 else {
        print("[MAC] ERROR: Session ID must be exactly 32 bytes, got \(sessionId.count)")
        return nil
    }

    let pathBytes = path.data(using: .utf8) ?? Data()
    guard pathBytes.count <= UInt16.max else {
        print("[MAC] ERROR: Path too long: \(pathBytes.count) bytes (max \(UInt16.max))")
        return nil
    }
    guard body.count <= UInt32.max else {
        print("[MAC] ERROR: Body too long: \(body.count) bytes (max \(UInt32.max))")
        return nil
    }

    var data = Data()

    // 1. Method (1 byte)
    data.append(methodByte)

    // 2. Path length (2 bytes BE) + path bytes
    var pathLen = UInt16(pathBytes.count).bigEndian
    data.append(Data(bytes: &pathLen, count: MemoryLayout<UInt16>.size))
    data.append(pathBytes)

    // 3. Body length (4 bytes BE) + body bytes
    var bodyLen = UInt32(body.count).bigEndian
    data.append(Data(bytes: &bodyLen, count: MemoryLayout<UInt32>.size))
    data.append(body)

    // 4. Counter (8 bytes BE)
    var counterBe = counter.bigEndian
    data.append(Data(bytes: &counterBe, count: MemoryLayout<UInt64>.size))

    // 5. Session ID (32 bytes)
    data.append(sessionId)

    return data
}

private func zhtpMethodByte(_ method: ZhtpMethod) -> UInt8? {
    switch method {
    case .get: return 0
    case .post: return 1
    case .put: return 2
    case .delete: return 3
    default: return nil
    }
}

// MARK: - MAC Computation (UHP v2)

func hkdfSha3(
    ikm: Data,
    salt: Data,
    info: Data,
    outputLength: Int
) throws -> Data {
    var output = Data(count: outputLength)
    let result = output.withUnsafeMutableBytes { outBuf -> Int32 in
        guard let outPtr = outBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
            return -1
        }
        return ikm.withUnsafeBytes { ikmBuf in
            guard let ikmPtr = ikmBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                return -1
            }
            return salt.withUnsafeBytes { saltBuf in
                let saltPtr = saltBuf.baseAddress?.assumingMemoryBound(to: UInt8.self)
                return info.withUnsafeBytes { infoBuf in
                    let infoPtr = infoBuf.baseAddress?.assumingMemoryBound(to: UInt8.self)
                    return uhp_hkdf_sha3_256(
                        ikmPtr,
                        ikm.count,
                        saltPtr,
                        salt.count,
                        infoPtr,
                        info.count,
                        outPtr,
                        outputLength
                    )
                }
            }
        }
    }

    if result != 0 {
        throw NSError(domain: "UHP", code: -1, userInfo: [NSLocalizedDescriptionKey: "HKDF failed"])
    }

    return output
}

func hmacSha3(key: Data, message: Data) throws -> Data {
    var output = Data(count: 32)
    let result = output.withUnsafeMutableBytes { outBuf -> Int32 in
        guard let outPtr = outBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
            return -1
        }
        return key.withUnsafeBytes { keyBuf in
            guard let keyPtr = keyBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                return -1
            }
            return message.withUnsafeBytes { msgBuf in
                guard let msgPtr = msgBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                    return -1
                }
                return uhp_hmac_sha3_256(
                    keyPtr,
                    key.count,
                    msgPtr,
                    message.count,
                    outPtr,
                    32
                )
            }
        }
    }

    if result != 0 {
        throw NSError(domain: "UHP", code: -1, userInfo: [NSLocalizedDescriptionKey: "HMAC failed"])
    }

    return output
}

/// Build AuthContext for a request (computes MAC)
///
/// Performs these steps:
/// 1. Get next monotonic sequence number
/// 2. Build canonical request bytes
/// 3. Compute HMAC-SHA3-256(macKey, canonicalRequest)
/// 4. Return AuthContext with sessionId, clientDid, sequence, and MAC
func buildAuthContext(
    session: inout AuthSession,
    method: ZhtpMethod,
    path: String,
    body: Data
) throws -> AuthContext {
    let sequence = session.nextSequence()

    guard let macInput = buildMacInput(
        method: method,
        path: path,
        body: body,
        counter: sequence,
        sessionId: session.sessionId
    ) else {
        throw NSError(domain: "ZHTP", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid MAC input"])
    }

    let requestMac = try hmacSha3(key: session.macKey, message: macInput)

    print("[Auth] Request MAC computed: sequence=\(sequence), macLen=\(requestMac.count), inputLen=\(macInput.count)")

    return AuthContext(
        session_id: session.sessionId,
        client_did: session.clientDid,
        sequence: sequence,
        request_mac: requestMac
    )
}

// MARK: - ZhtpMethod Wire Encoding

extension ZhtpMethod {
    var wireValue: Int {
        switch self {
        case .get: return 0
        case .post: return 1
        case .put: return 2
        case .delete: return 3
        case .options: return 4
        case .head: return 5
        case .patch: return 6
        case .verify: return 7
        case .connect: return 8
        case .trace: return 9
        }
    }
}

// MARK: - UHP Identity Keystore Loader

struct UhpPrivateKeyBytesData {
    let dilithiumSk: Data
    let kyberSk: Data
    let masterSeed: Data
}

struct UhpIdentityMaterials {
    let identityJson: Data
    let identityDid: String
    let privateKey: UhpPrivateKeyBytesData
}

private struct UhpPrivateKeyJson: Decodable {
    let dilithium_sk: [UInt8]
    let kyber_sk: [UInt8]
    let master_seed: [UInt8]
}

enum UhpKeystore {
    private static let keychainService = "com.sovereign.zhtp"
    private static let keychainAccountPrefix = "private_key_"

    /// Store identity materials when synced from server
    static func storeIdentityMaterials(identityId: String, identityJson: Data, privateKeyJson: Data) -> Bool {
        guard let documentsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            print("[UhpKeystore] Failed to get Documents directory")
            return false
        }

        let keystoreDir = documentsDir.appendingPathComponent("keystore")
        let identityDir = keystoreDir.appendingPathComponent(identityId)

        do {
            // Create directories if they don't exist
            try FileManager.default.createDirectory(at: identityDir, withIntermediateDirectories: true)

            // Store identity JSON file
            let identityPath = identityDir.appendingPathComponent("user_identity.json")
            try identityJson.write(to: identityPath)
            print("[UhpKeystore] ✅ Stored identity JSON at: \(identityPath)")

            // Store private key in Keychain
            let success = storePrivateKeyJson(identityId: identityId, data: privateKeyJson, requireUserPresence: true)
            if success {
                print("[UhpKeystore] ✅ Stored private key in Keychain")
                return true
            } else {
                print("[UhpKeystore] ❌ Failed to store private key in Keychain")
                return false
            }
        } catch {
            print("[UhpKeystore] ❌ Failed to store identity: \(error)")
            return false
        }
    }

    static func loadIdentityForHandshake(identityId: String) -> UhpIdentityMaterials? {
        let trimmedId = identityId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedId.isEmpty else {
            print("[UhpKeystore] ❌ Identity ID is empty")
            return nil
        }

        guard let identityJson = loadIdentityJson(identityId: trimmedId) else {
            print("[UhpKeystore] ❌ Failed to load identity JSON from Documents/keystore/\(trimmedId)/user_identity.json")
            return nil
        }
        print("[UhpKeystore] ✅ Loaded identity JSON (\(identityJson.count) bytes)")

        guard let identityDid = parseIdentityDid(identityJson) else {
            print("[UhpKeystore] ❌ Failed to parse DID from identity JSON")
            return nil
        }
        print("[UhpKeystore] ✅ Parsed identity DID: \(identityDid)")

        guard let privateKeyJson = loadPrivateKeyJson(identityId: trimmedId) else {
            print("[UhpKeystore] ❌ Failed to load private key from Keychain with account 'private_key_\(trimmedId)'")
            return nil
        }
        print("[UhpKeystore] ✅ Loaded private key JSON from Keychain (\(privateKeyJson.count) bytes)")

        guard let privateKey = decodePrivateKey(privateKeyJson) else {
            print("[UhpKeystore] ❌ Failed to decode private key JSON (missing or invalid dilithium_sk, kyber_sk, master_seed)")
            return nil
        }
        print("[UhpKeystore] ✅ Decoded private key: Dilithium(\(privateKey.dilithiumSk.count)) Kyber(\(privateKey.kyberSk.count)) Seed(\(privateKey.masterSeed.count))")

        return UhpIdentityMaterials(identityJson: identityJson, identityDid: identityDid, privateKey: privateKey)
    }

    private static func loadIdentityJson(identityId: String) -> Data? {
        guard let documentsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            print("[UhpKeystore] Failed to get Documents directory")
            return nil
        }

        let keystoreDir = documentsDir.appendingPathComponent("keystore")
        let identityDir = keystoreDir.appendingPathComponent(identityId)
        let identityPath = identityDir.appendingPathComponent("user_identity.json")

        do {
            return try Data(contentsOf: identityPath)
        } catch {
            print("[UhpKeystore] ⚠️ Identity file missing at: \(identityPath)")
            print("[UhpKeystore] This identity needs to be synced from the server or created.")
            print("[UhpKeystore] Expected identity materials:")
            print("[UhpKeystore]   - Dilithium5 signing keypair (for UHP v2 ClientHello signature)")
            print("[UhpKeystore]   - Kyber1024 KEM keypair (for post-quantum key exchange)")
            print("[UhpKeystore]   - Identity DID, public keys, node ID, device ID, timestamp")
            print("[UhpKeystore]")
            print("[UhpKeystore] Next steps:")
            print("[UhpKeystore]   1. Fetch identity from server: GET /api/v1/identities/\(identityId)")
            print("[UhpKeystore]   2. Store response at: \(identityPath)")
            print("[UhpKeystore]   3. Also store key materials in Keychain with account: private_key_\(identityId)")
            return nil
        }
    }

    static func storePrivateKeyJson(identityId: String, data: Data, requireUserPresence: Bool) -> Bool {
        let account = "\(keychainAccountPrefix)\(identityId)"
        guard let access = createAccessControl(requireUserPresence: requireUserPresence) else { return false }

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: account,
        ]

        let attributes: [String: Any] = [
            kSecValueData as String: data,
            kSecAttrAccessControl as String: access,
        ]

        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if status == errSecSuccess {
            return true
        }

        var addQuery = query
        addQuery[kSecValueData as String] = data
        addQuery[kSecAttrAccessControl as String] = access
        addQuery[kSecUseAuthenticationUI as String] = kSecUseAuthenticationUIAllow
        let addStatus = SecItemAdd(addQuery as CFDictionary, nil)
        return addStatus == errSecSuccess
    }

    private static func loadPrivateKeyJson(identityId: String) -> Data? {
        let account = "\(keychainAccountPrefix)\(identityId)"
        var query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]

        query[kSecUseOperationPrompt as String] = "Authenticate to access your private key"

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess else {
            let errorMsg: String
            switch status {
            case errSecItemNotFound:
                errorMsg = "Item not found in Keychain"
            case errSecAuthFailed:
                errorMsg = "Authentication failed (user denied)"
            case errSecUserCanceled:
                errorMsg = "User cancelled authentication"
            default:
                errorMsg = "OSStatus \(status)"
            }
            print("[UhpKeystore] Keychain lookup failed for '\(account)': \(errorMsg)")
            return nil
        }

        guard let data = item as? Data else {
            print("[UhpKeystore] Keychain item is not Data type")
            return nil
        }

        return data
    }

    private static func decodePrivateKey(_ data: Data) -> UhpPrivateKeyBytesData? {
        do {
            let decoded = try JSONDecoder().decode(UhpPrivateKeyJson.self, from: data)
            return UhpPrivateKeyBytesData(
                dilithiumSk: Data(decoded.dilithium_sk),
                kyberSk: Data(decoded.kyber_sk),
                masterSeed: Data(decoded.master_seed)
            )
        } catch {
            print("[UhpKeystore] JSON decode error: \(error)")
            if let jsonStr = String(data: data, encoding: .utf8) {
                print("[UhpKeystore] JSON content: \(jsonStr.prefix(200))...")
            }
            return nil
        }
    }

    private static func parseIdentityDid(_ identityJson: Data) -> String? {
        guard
            let obj = try? JSONSerialization.jsonObject(with: identityJson),
            let dict = obj as? [String: Any],
            let did = dict["did"] as? String,
            !did.isEmpty
        else {
            return nil
        }
        return did
    }

    private static func createAccessControl(requireUserPresence: Bool) -> SecAccessControl? {
        let flags: SecAccessControlCreateFlags = requireUserPresence ? .userPresence : []
        return SecAccessControlCreateWithFlags(
            nil,
            kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            flags,
            nil
        )
    }
}

File: android/app/src/main/rust/quic-jni/src/zhtp_auth.rs
//! ZHTP Authenticated Mode - UHP Handshake + Session Management
//!
//! Implements authenticated requests with Dilithium5 signing and Kyber512 key exchange.

use crate::zhtp_types::{ZhtpMethod, ZhtpHeaders};
use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};

/// Session state for authenticated connections
#[derive(Debug, Clone)]
pub struct AuthSession {
    pub session_id: Vec<u8>,      // [u8; 16] from handshake
    pub app_key: Vec<u8>,          // [u8; 32] from HKDF derivation
    pub sequence: u64,             // monotonic counter, incremented per request
    pub client_did: String,        // client identity
    pub server_did: String,        // server identity
    pub created_at: u64,           // unix timestamp in seconds
    pub last_activity: u64,        // unix timestamp in seconds
}

impl AuthSession {
    /// Check if session is still valid
    /// - Not idle for > 5 minutes
    /// - Not older than 1 hour
    pub fn is_valid(&self) -> bool {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();

        // 5-minute idle timeout
        if now.saturating_sub(self.last_activity) > 300 {
            return false;
        }

        // 1-hour age limit
        if now.saturating_sub(self.created_at) > 3600 {
            return false;
        }

        true
    }

    /// Update last activity timestamp
    pub fn touch(&mut self) {
        self.last_activity = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();
    }

    /// Increment sequence counter for next request
    pub fn next_sequence(&mut self) -> u64 {
        let current = self.sequence;
        self.sequence = self.sequence.saturating_add(1);
        current
    }
}

/// Authentication context sent in ZHTP request header
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthContext {
    pub session_id: Vec<u8>,  // [u8; 16]
    pub client_did: String,
    pub sequence: u64,
    pub request_mac: Vec<u8>, // [u8; 32]
}

/// Compute canonical hash of request following exact server order
/// Hash input:
/// 1. WIRE_VERSION (u16 LE) = 1
/// 2. request_id (16 bytes)
/// 3. timestamp_ms (u64 LE)
/// 4. method encoded (1 byte)
/// 5. uri (length-prefixed string)
/// 6. headers in fixed order (content_type, content_length, content_encoding, cache_control)
/// 7. body (length-prefixed bytes)
pub fn compute_canonical_request_hash(
    request_id: &[u8; 16],
    timestamp_ms: u64,
    method: ZhtpMethod,
    uri: &str,
    headers: &ZhtpHeaders,
    body: &[u8],
) -> Vec<u8> {
    let mut hash_input = Vec::new();

    // 1. WIRE_VERSION (u16 LE) = 1
    hash_input.extend_from_slice(&(1u16).to_le_bytes());

    // 2. request_id (16 bytes)
    hash_input.extend_from_slice(request_id);

    // 3. timestamp_ms (u64 LE)
    hash_input.extend_from_slice(&timestamp_ms.to_le_bytes());

    // 4. method encoded (1 byte)
    let method_byte = match method {
        ZhtpMethod::Get => 0,
        ZhtpMethod::Post => 1,
        ZhtpMethod::Put => 2,
        ZhtpMethod::Delete => 3,
        ZhtpMethod::Options => 4,
        ZhtpMethod::Head => 5,
        ZhtpMethod::Patch => 6,
        ZhtpMethod::Verify => 7,
        ZhtpMethod::Connect => 8,
        ZhtpMethod::Trace => 9,
    };
    hash_input.push(method_byte);

    // 5. uri (length-prefixed string)
    let uri_bytes = uri.as_bytes();
    hash_input.extend_from_slice(&(uri_bytes.len() as u32).to_le_bytes());
    hash_input.extend_from_slice(uri_bytes);

    // 6. headers in fixed order
    // content_type (present flag + length + bytes)
    if let Some(ct) = &headers.content_type {
        hash_input.push(1); // present flag
        let ct_bytes = ct.as_bytes();
        hash_input.extend_from_slice(&(ct_bytes.len() as u32).to_le_bytes());
        hash_input.extend_from_slice(ct_bytes);
    } else {
        hash_input.push(0); // not present
    }

    // content_length (present flag + u64 LE)
    if let Some(cl) = headers.content_length {
        hash_input.push(1); // present flag
        hash_input.extend_from_slice(&cl.to_le_bytes());
    } else {
        hash_input.push(0);
    }

    // content_encoding (present flag + length + bytes)
    if let Some(ce) = &headers.content_encoding {
        hash_input.push(1);
        let ce_bytes = ce.as_bytes();
        hash_input.extend_from_slice(&(ce_bytes.len() as u32).to_le_bytes());
        hash_input.extend_from_slice(ce_bytes);
    } else {
        hash_input.push(0);
    }

    // cache_control (present flag + length + bytes)
    if let Some(cc) = &headers.cache_control {
        hash_input.push(1);
        let cc_bytes = cc.as_bytes();
        hash_input.extend_from_slice(&(cc_bytes.len() as u32).to_le_bytes());
        hash_input.extend_from_slice(cc_bytes);
    } else {
        hash_input.push(0);
    }

    // 7. body (length-prefixed bytes)
    hash_input.extend_from_slice(&(body.len() as u32).to_le_bytes());
    hash_input.extend_from_slice(body);

    // Hash with BLAKE3
    blake3::hash(&hash_input).as_bytes().to_vec()
}

/// Compute request MAC using BLAKE3 keyed hashing
/// MAC = BLAKE3_keyed(app_key, session_id || sequence || canonical_hash)
pub fn compute_request_mac(
    app_key: &[u8],
    session_id: &[u8],
    sequence: u64,
    canonical_hash: &[u8],
) -> Vec<u8> {
    let mut mac_input = Vec::new();
    mac_input.extend_from_slice(session_id);
    mac_input.extend_from_slice(&sequence.to_le_bytes());
    mac_input.extend_from_slice(canonical_hash);

    // Derive keyed MAC from app_key
    let key_hash = blake3::hash(app_key);
    let key_array: [u8; 32] = key_hash.as_bytes()[..32].try_into().unwrap_or_default();

    blake3::keyed_hash(&key_array, &mac_input)
        .as_bytes()
        .to_vec()
}

/// Derive app_key from master_key using BLAKE3
/// app_key = blake3("zhtp-web4-app-mac" || master_key || session_id || server_did || client_did)
pub fn derive_app_key(
    master_key: &[u8],
    session_id: &[u8],
    server_did: &str,
    client_did: &str,
) -> Vec<u8> {
    let mut input = Vec::new();
    input.extend_from_slice(b"zhtp-web4-app-mac");
    input.extend_from_slice(master_key);
    input.extend_from_slice(session_id);
    input.extend_from_slice(server_did.as_bytes());
    input.extend_from_slice(client_did.as_bytes());

    blake3::hash(&input).as_bytes().to_vec()
}

/// Derive master key from session components using BLAKE3
/// Master Key = BLAKE3(
///   "zhtp-quic-master" ||
///   uhp_session_key || pqc_shared_secret || uhp_transcript_hash || peer_node_id
/// )
pub fn derive_master_key(
    uhp_session_key: &[u8],
    pqc_shared_secret: &[u8],
    uhp_transcript_hash: &[u8],
    peer_node_id: &str,
) -> Vec<u8> {
    let mut input = Vec::new();
    input.extend_from_slice(b"zhtp-quic-master");
    input.extend_from_slice(uhp_session_key);
    input.extend_from_slice(pqc_shared_secret);
    input.extend_from_slice(uhp_transcript_hash);
    input.extend_from_slice(peer_node_id.as_bytes());

    blake3::hash(&input).as_bytes()[..32].to_vec()
}

/// Build AuthContext for a request
pub fn build_auth_context(
    session: &AuthSession,
    canonical_hash: &[u8],
) -> Result<AuthContext> {
    let mut session_mut = session.clone();
    let sequence = session_mut.next_sequence();

    let request_mac = compute_request_mac(&session.app_key, &session.session_id, sequence, canonical_hash);

    Ok(AuthContext {
        session_id: session.session_id.clone(),
        client_did: session.client_did.clone(),
        sequence,
        request_mac,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_session_validity() {
        let session = AuthSession {
            session_id: vec![0u8; 16],
            app_key: vec![0u8; 32],
            sequence: 0,
            client_did: "client".to_string(),
            server_did: "server".to_string(),
            created_at: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs(),
            last_activity: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        };
        assert!(session.is_valid());
    }

    #[test]
    fn test_canonical_hash_deterministic() {
        let request_id = [1u8; 16];
        let method = ZhtpMethod::Post;
        let uri = "/api/v1/transactions";
        let headers = ZhtpHeaders {
            content_type: Some("application/json".to_string()),
            content_length: Some(100),
            dao_fee: 10,
            total_fees: 11,
            content_encoding: None,
            cache_control: None,
            network_fee: None,
            priority: None,
        };
        let body = b"test";

        let hash1 = compute_canonical_request_hash(&request_id, 1000, method, uri, &headers, body);
        let hash2 = compute_canonical_request_hash(&request_id, 1000, method, uri, &headers, body);

        assert_eq!(hash1, hash2, "Canonical hash must be deterministic");
        assert_eq!(hash1.len(), 32, "BLAKE3 hash should be 32 bytes");
    }
}
